### PR TITLE
Orga mediafiles

### PR DESCRIPTION
**Pull request by reiterl**

### CODE REVIEW

**Comment by reiterl** [comment]
Draft version, some stuff is missing. WIP.

**Comment by reiterl** [comment]
I have removed the resource collection and update the tests and code. Also I have add owner_id as generic field. It is a rename of the meeting_id and add the token field. I have added two migrations with tests. One removed the resource, the other updated the  meeting_id. This PR should not need changes on the media service layer.

File: openslides_backend/action/actions/mediafile/upload.py [hunk 1]
```diff
@@ -31,23 +28,19 @@
...
+        required_properties=["title", "owner_id", "filename"],
         optional_properties=["access_group_ids", "parent_id"],
         additional_required_fields={"file": {"type": "string"}},
     )
     permission = Permissions.Mediafile.CAN_MANAGE
-
-    dependencies = [ListOfSpeakersCreate]
+    skip_archived_meeting_check = True
 
     def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/upload.py:45]: Here, the code from the old `resource.upload` has to be included: if `token` is given, `pop` the `filename` from the instance. Also the `get_updated_instance` method from the old action has to be included.

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/mediafile/create_directory.py [hunk 2]
```diff
@@ -7,20 +7,22 @@
...
     """
 
     model = Mediafile()
     schema = DefaultSchema(Mediafile()).get_create_schema(
-        required_properties=["meeting_id", "title"],
+        required_properties=["owner_id", "title"],
         optional_properties=["access_group_ids", "parent_id"],
     )
     permission = Permissions.Mediafile.CAN_MANAGE
+    skip_archived_meeting_check = True
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/create_directory.py]: I overwrite `check_for_archived_meeting` and add a helper get_owner_data. I remove the `skip_check_for_archived_meeting`.

File: openslides_backend/action/actions/mediafile/delete.py [hunk 3]
```diff
@@ -7,17 +7,19 @@
...
-class MediafileDelete(DeleteAction):
+class MediafileDelete(MediafilePermissionMixin, DeleteAction):
     """
     Action to delete a user.
     """
 
     model = Mediafile()
     schema = DefaultSchema(Mediafile()).get_delete_schema()
     permission = Permissions.Mediafile.CAN_MANAGE
+    skip_archived_meeting_check = True
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/delete.py]: Removed.

File: openslides_backend/action/actions/mediafile/permission_mixin.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        # check for archived meeting
+        fqid = FullQualifiedId(Collection("meeting"), id_)
+        meeting = self.datastore.fetch_model(
+            fqid,
+            ["is_active_in_organization_id", "name"],
+        )
+        if not meeting.get("is_active_in_organization_id"):
+            raise ActionException(
+                f'Meeting {meeting.get("name", "")}/{id_} cannot be changed, because it is archived.'
+            )
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/permission_mixin.py]: Use the overwritten check_for_archived_meeting, which calls the super function, if needed.

File: openslides_backend/action/actions/mediafile/update.py [hunk 5]
```diff
@@ -20,6 +23,7 @@ class MediafileUpdate(UpdateAction, MediafileCalculatedFieldsMixin):
         optional_properties=["title", "access_group_ids"]
     )
     permission = Permissions.Mediafile.CAN_MANAGE
+    skip_archived_meeting_check = True
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/update.py]: Removed.

File: openslides_backend/action/actions/mediafile/upload.py [hunk 6]
```diff
@@ -31,23 +28,19 @@
...
     schema = DefaultSchema(Mediafile()).get_create_schema(
-        required_properties=["title", "meeting_id", "filename"],
+        required_properties=["title", "owner_id", "filename"],
         optional_properties=["access_group_ids", "parent_id"],
         additional_required_fields={"file": {"type": "string"}},
     )
     permission = Permissions.Mediafile.CAN_MANAGE
-
-    dependencies = [ListOfSpeakersCreate]
+    skip_archived_meeting_check = True
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/upload.py]: Removed.

File: openslides_backend/action/actions/mediafile/upload.py [hunk 7]
```diff
@@ -31,23 +28,19 @@
...
+class MediafileUploadAction(MediafilePermissionMixin, CreateAction):
     """
     Action to upload a mediafile.
     """
 
     model = Mediafile()
     schema = DefaultSchema(Mediafile()).get_create_schema(
-        required_properties=["title", "meeting_id", "filename"],
+        required_properties=["title", "owner_id", "filename"],
         optional_properties=["access_group_ids", "parent_id"],
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/upload.py:39]: Added token to optional_properties.

File: global/meta/models.yml [hunk 8]
```diff
@@ -3025,30 +3003,25 @@ mediafile:
   access_group_ids:
     type: relation-list
     to: group/mediafile_access_group_ids
-    equal_fields: meeting_id
     restriction_mode: A
   parent_id:
     type: relation
     to: mediafile/child_ids
-    equal_fields: meeting_id
```
[the remark this answers is not in the document]
Comment (reiterl) [reply global/meta/models.yml:3025]: Readded with owner_id.

File: global/meta/models.yml [hunk 9]
```diff
@@ -3025,30 +3003,25 @@ mediafile:
...
     restriction_mode: A
     required: false
   child_ids:
     type: relation-list
     to: mediafile/parent_id
-    equal_fields: meeting_id
```
[the remark this answers is not in the document]
Comment (reiterl) [reply global/meta/models.yml:3031]: Readded with owner_id.

File: openslides_backend/presenter/check_resource_id.py [hunk 10]
```diff
@@ -1,47 +0,0 @@
...
-
-@register_presenter("check_resource_id")
-class CheckResourceId(BasePresenter):
-    """
-    Check, if a resource can be accessed. Retrieve the filename, if access is granted.
-    """
-
-    schema = check_resource_id_schema
-
-    def get_result(self) -> Any:
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/presenter/check_resource_id.py:33]: Move the resource get_result code into the check_mediafile_id presenter. Build in a switch which depends of the of the owner_id.

**Comment by reiterl** [comment]
I wait with more changes on the documentation of the upload action.

File: tests/system/presenter/test_check_mediafile_id.py [hunk 11]
```diff
@@ -158,3 +166,60 @@ def test_can_see_and_inherited_groups(self) -> None:
         )
         status_code, data = self.request("check_mediafile_id", {"mediafile_id": 1})
         self.assertEqual(status_code, 200)
+
+    def test_simple_organization(self) -> None:
```
[the remark this answers is not in the document]
Comment (reiterl) [reply tests/system/presenter/test_check_mediafile_id.py:175]: Update the organization owner tests.

File: openslides_backend/action/actions/mediafile/upload.py (hunk 1 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/upload.py:75]: Merge the `resource.upload` action into the `mediafile.upload` action.

File: tests/system/presenter/test_check_mediafile_id.py [hunk 12]
```diff
@@ -158,3 +166,60 @@ def test_can_see_and_inherited_groups(self) -> None:
...
+        )
+
+        response = self.anon_client.post(
+            PRESENTER_URL,
+            json=[{"presenter": "check_mediafile_id", "data": {"mediafile_id": 1}}],
+        )
+        status_code = response.status_code
+        self.assertEqual(status_code, 403)
+
+    def test_anonymous_organization_with_token(self) -> None:
```
[the remark this answers is not in the document]
Comment (reiterl) [reply tests/system/presenter/test_check_mediafile_id.py:227]: Updated.

File: openslides_backend/action/actions/resource/upload.py [hunk 13]
```diff
@@ -1,75 +0,0 @@
...
-            raise ActionException(f"Cannot guess mimetype for {filename_}.")
-        decoded_file = base64.b64decode(file_)
-        instance["filesize"] = len(decoded_file)
-        id_ = instance["id"]
-        mimetype_ = instance["mimetype"]
-        self.media.upload_resource(file_, id_, mimetype_)
-        return instance
-
-    @original_instances
-    def get_updated_instances(self, action_data: ActionData) -> ActionData:
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/resource/upload.py:50]: Done.

File: openslides_backend/action/actions/meeting/base_set_mediafile_action.py [hunk 14]
```diff
@@ -52,3 +51,11 @@ def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
...
         instance[self.field] = {instance.pop("place"): instance.pop("mediafile_id")}
         return instance
+
+    def check_owner(self, mediafile: Dict[str, Any], instance: Dict[str, Any]) -> None:
+        owner_id = mediafile["owner_id"]
+        collection, id_ = owner_id.split(KEYSEPARATOR)
+        if collection != "meeting":
+            raise ActionException("Mediafile is not a meeting mediafile.")
+        if int(id_) != instance["id"]:
+            raise ActionException("Mediafile does not belong to the meeting.")
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/meeting/base_set_mediafile_action.py:61]: Reuse of the old error message.

File: openslides_backend/action/util/assert_belongs_to_meeting.py [hunk 15]
```diff
@@ -26,6 +26,8 @@ def assert_belongs_to_meeting(
             )
             if meeting_id not in instance.get("meeting_ids", []):
                 errors.add(str(fqid))
+        elif fqid.collection.collection == "mediafile":
+            pass
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/util/assert_belongs_to_meeting.py:30]: Add a check for the meeting_id, if the mediafile has a meeting owner.

**Comment by reiterl** [comment]
Rebased.
Update check_mediafile_id and  mediafile.update.

File: openslides_backend/action/actions/mediafile/permission_mixin.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    ActionException,
+    MissingPermission,
+    PermissionException,
+)
+from ....shared.filters import And, Filter, FilterOperator, Not
+from ....shared.patterns import KEYSEPARATOR, Collection, FullQualifiedId
+from ...action import Action
+
+
+class MediafilePermissionMixin(Action):
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/permission_mixin.py:16]: Thisis now more of a general mediafile mixin instead of a permission mixin, so please rename

File: openslides_backend/action/actions/mediafile/permission_mixin.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+    """
+    Mixin to handle the check_permissions of mediafile actions.
+    Overwrite check_permissions() and get_meeting_id().
+    """
+
+    def check_permissions(self, instance: Dict[str, Any]) -> None:
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/permission_mixin.py:22]: This does way more than checking permissions. Everythign which doesn't throw a PermissionException should not be part of the `check_permissions` method, but of e.g. `update_instance`. Please move this accordingly and call `super().update_instance()` in the child classes, where necessary.

File: openslides_backend/action/actions/mediafile/permission_mixin.py [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+        self.check_title_parent_unique(
+            instance.get("title"), parent_id, instance.get("id")
+        )
+
+        # handle organization permissions
+        if collection == "organization":
+            self.assert_not_anonymous()
+            self.check_token_unique(instance.get("token"), instance.get("id"))
+            if "access_group_ids" in instance:
+                raise PermissionException(
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/permission_mixin.py:43]: This should be an `ActionException`, since the wrong payload was given. `PermissionError`s are for situations where the user has insufficent permissions to execute the given permissions, but otherwise the request is correct.

File: openslides_backend/action/actions/mediafile/permission_mixin.py [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+                raise MissingPermission(
+                    OrganizationManagementLevel.CAN_MANAGE_ORGANIZATION
+                )
+            return
+
+        assert collection == "meeting"
+
+        # check for token, not allowed in meeting.
+        if "token" in instance:
+            raise PermissionException("token is not allowed in meeting mediafiles.")
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/permission_mixin.py:60]: This should be an `ActionException`, see above

File: openslides_backend/action/actions/mediafile/upload.py [hunk 20]
```diff
@@ -31,25 +32,71 @@
...
+    def get_updated_instances(self, action_data: ActionData) -> ActionData:
+        tokens: List[Any] = []
+        for instance in action_data:
+            collection, _ = self.get_owner_data(instance)
+            if collection != "organization":
+                continue
+            tokens.append(instance.get("token"))
+            results = self.datastore.filter(
+                self.model.collection,
+                FilterOperator("token", "=", instance["token"]),
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/upload.py:58]: also add the owner_id filter here (see above)

File: openslides_backend/action/actions/mediafile/upload.py [hunk 21]
```diff
@@ -31,25 +32,71 @@
...
+        file_ = instance.pop("file")
+        instance["mimetype"] = mimetypes.guess_type(filename_)[0]
+        if instance["mimetype"] is None:
+            raise ActionException(f"Cannot guess mimetype for {filename_}.")
+        decoded_file = base64.b64decode(file_)
+        instance["filesize"] = len(decoded_file)
+        id_ = instance["id"]
+        mimetype_ = instance["mimetype"]
+        self.media.upload_mediafile(file_, id_, mimetype_)
+        return instance
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/upload.py:97]: You can combine this with the `update_meeting_instance`: Everything done here should also be done for the meeting case. `pop`ing the `filename` is not necessary and `pop`ing the `file` itself must also be done for the meeting case - it seems like you do not do this atm. Please also add a test for this if that is the case.

File: openslides_backend/presenter/check_mediafile_id.py [hunk 22]
```diff
@@ -38,30 +43,71 @@ class CheckMediafileId(BasePresenter):
...
                 "inherited_access_group_ids",
+                "token",
             ],
         )
-        meeting_id = mediafile.get("meeting_id")
+
+        # Try to get the meeting id.
+        meeting_id = None
+        if mediafile.get("owner_id"):
+            if mediafile["owner_id"].split(KEYSEPARATOR)[0] == "organization":
```
Comment (jsangmeister) [inline openslides_backend/presenter/check_mediafile_id.py:104]: save result of `split` for easier access

File: openslides_backend/presenter/check_mediafile_id.py [hunk 23]
```diff
@@ -38,30 +43,71 @@ class CheckMediafileId(BasePresenter):
...
+        if extension is None:
+            return {"ok": False}
+        filename = mediafile["token"] + extension
+
+        return {"ok": True, "filename": filename}
 
+    def get_meeting_result(self, mediafile: Dict[str, Any]) -> Any:
+        if not mediafile.get("filename"):
+            return {"ok": False}
         return {"ok": True, "filename": mediafile["filename"]}
```
Comment (jsangmeister) [inline openslides_backend/presenter/check_mediafile_id.py:85]: Check the description in https://github.com/OpenSlides/OpenSlides/wiki/check_mediafile_id again: You should be able to unify these methods, the only exception is if a token is set. Also, you are currently calling `check_permissions` twice to organization mediafiles.

File: tests/system/action/mediafile/test_create_directory.py [hunk 24]
```diff
@@ -315,25 +328,139 @@ def test_create_directory_no_access_group_ids(self) -> None:
...
+            "mediafile.create_directory",
+            {
+                "owner_id": "meeting/1",
+                "title": "title_1",
+                "parent_id": 6,
+            },
+        )
+        self.assert_status_code(response, 400)
+        assert (
+            "Title 'title_1' and parent_id '6' are not unique."
```
Comment (jsangmeister) [inline tests/system/action/mediafile/test_create_directory.py:401]: Please change this error message, e.g. "File x already exists in folder y" (and fetch the name of the folder to replace y)

File: tests/system/action/mediafile/test_create_directory.py [hunk 25]
```diff
@@ -315,25 +328,139 @@ def test_create_directory_no_access_group_ids(self) -> None:
...
+        response = self.request(
+            "mediafile.create_directory",
+            {
+                "owner_id": "meeting/2",
+                "title": "title_1",
+                "parent_id": 7,
+            },
+        )
+        self.assert_status_code(response, 400)
+        assert "Owner and parent don't match." in response.json["message"]
```
Comment (jsangmeister) [inline tests/system/action/mediafile/test_create_directory.py:370]: Add a similar test that you cannot mix organization and meeting mediafiles.

File: tests/system/action/meeting/test_import.py [hunk 26]
```diff
@@ -301,9 +301,10 @@ def get_mediafile_data(
         file_content = base64.b64encode(b"testtesttest").decode()
         return {
             "id": obj_id,
-            "meeting_id": 1,
+            "owner_id": "meeting/1",
             "blob": file_content,
             "title": "A.txt",
+            "token": None,
```
Comment (jsangmeister) [inline tests/system/action/meeting/test_import.py:307]: remove

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: tests/system/action/meeting/test_import.py (hunk 26 above)
Comment (reiterl) [reply tests/system/action/meeting/test_import.py]: Removed.

File: tests/system/action/mediafile/test_create_directory.py (hunk 25 above)
Comment (reiterl) [reply tests/system/action/mediafile/test_create_directory.py:370]: Added.

File: tests/system/action/mediafile/test_create_directory.py (hunk 24 above)
Comment (reiterl) [reply tests/system/action/mediafile/test_create_directory.py]: Renamed into  f"File '{title}' already exists in folder '{parent_title}'."
And updated the tests.

File: openslides_backend/presenter/check_mediafile_id.py (hunk 23 above)
Comment (reiterl) [reply openslides_backend/presenter/check_mediafile_id.py:85]: Ah, okay, I have unified the two methods.

File: openslides_backend/presenter/check_mediafile_id.py (hunk 22 above)
Comment (reiterl) [reply openslides_backend/presenter/check_mediafile_id.py]: Improved the handling of the owner_collection and the owner_id_int.
The name owner_id_int is strange, but I think owner_id_id is worse.

File: openslides_backend/action/actions/mediafile/permission_mixin.py (hunk 19 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/permission_mixin.py:60]: Changed into ActionException.

File: openslides_backend/action/actions/mediafile/permission_mixin.py (hunk 18 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/permission_mixin.py:43]: Changed.

File: openslides_backend/action/actions/mediafile/create_directory.py [hunk 27]
```diff
@@ -28,21 +30,26 @@ def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
...
         """
         instance["is_directory"] = True
-        if instance.get("parent_id") is not None:
-            parent = self.datastore.get(
-                FullQualifiedId(self.model.collection, instance["parent_id"]),
-                ["is_public", "inherited_access_group_ids"],
-            )
+        instance["create_timestamp"] = round(time.time())
+        collection, id_ = self.get_owner_data(instance)
+        if collection == "meeting":
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/create_directory.py:35]: It is forbidden in the check_permission, no need to forbid it here.

File: migrations/tests/test_0019_remove_resource.py [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+            "fqid": "organization/1",
+            "fields": {"resource_ids": None},
+        },
+    )
+
+    finalize("0019_remove_resource")
+    assert_model(
+        "organization/1",
+        {"id": 1, "meta_deleted": False, "meta_position": 3},
+        position=3,
```
[the remark this answers is not in the document]
Comment (reiterl) [reply migrations/tests/test_0019_remove_resource.py:62]: Add asserts for all positions in the 0019 test.

[the remark this answers is not in the document]
Comment (reiterl) [reply migrations/tests/test_0019_remove_resource.py:62]: Added asserts for all positions in the remove_resource (0019) test.

File: migrations/tests/test_0020_update_mediafile_meeting_id.py [hunk 29]
```diff
@@ -0,0 +1,36 @@
+def test_migration(write, finalize, assert_model):
```
[the remark this answers is not in the document]
Comment (reiterl) [reply migrations/tests/test_0020_update_mediafile_meeting_id.py:1]: Added some more events with different event types to the test. Added asserts.

File: openslides_backend/action/actions/mediafile/permission_mixin.py [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+            )
+            gm_result = self.datastore.get_many([gm_request])
+            groups = gm_result.get(Collection("group"), {}).values()
+            for group in groups:
+                if group.get("meeting_id") != meeting_id:
+                    raise ActionException("Owner and access groups don't match.")
+
+    def check_token_unique(self, token: Optional[str], id_: Optional[int]) -> None:
+        if token:
+            filter_: Filter = FilterOperator("token", "=", token)
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/permission_mixin.py]: Added.

File: openslides_backend/action/actions/mediafile/permission_mixin.py (hunk 16 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/permission_mixin.py]: Renamed into `MediafileMixin`. I have also moved the mixin into a mixins file.

File: openslides_backend/action/actions/mediafile/permission_mixin.py (hunk 17 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/permission_mixin.py]: I have moved the check code which throws ActionExceptions into a update_instance method of the mixin.

File: openslides_backend/action/actions/mediafile/upload.py (hunk 20 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/upload.py]: Done.

File: openslides_backend/action/actions/mediafile/upload.py (hunk 21 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/upload.py]: Merge update_organization_instance and update_meeting_instance. Modify the code. File needs be popped and was popped in both methods.

File: openslides_backend/action/actions/mediafile/create_directory.py [hunk 31]
```diff
@@ -28,21 +30,26 @@ def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
...
+                ) = calculate_inherited_groups_helper(
+                    instance.get("access_group_ids"),
+                    parent.get("is_public"),
+                    parent.get("inherited_access_group_ids"),
+                )
+            else:
+                instance["inherited_access_group_ids"] = instance.get(
+                    "access_group_ids"
+                )
+                instance["is_public"] = not bool(instance["inherited_access_group_ids"])
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/create_directory.py:55]: Waiting for https://github.com/OpenSlides/openslides-backend/pull/1231

File: openslides_backend/action/actions/mediafile/permission_mixin.py [hunk 32]
```diff
@@ -0,0 +1,139 @@
...
+    def check_title_parent_unique(
+        self, title: Optional[str], parent_id: Optional[str], id_: Optional[int]
+    ) -> None:
+        if title:
+            filter_ = And(
+                FilterOperator("title", "=", title),
+                FilterOperator("parent_id", "=", parent_id),
+            )
+            if id_:
+                filter_ = And(filter_, Not(FilterOperator("id", "=", id_)))
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/permission_mixin.py]: Later.

File: migrations/tests/test_0020_update_mediafile_meeting_id.py [hunk 33]
```diff
@@ -11,6 +11,37 @@ def test_migration(write, finalize, assert_model):
...
+            "type": "delete",
+            "fqid": "mediafile/13",
+            "fields": {},
+        },
+        {
+            "type": "update",
+            "fqid": "meeting/8",
+            "fields": {"mediafile_ids": [12]},
+        },
+    )
```
Comment (jsangmeister) [inline migrations/tests/test_0020_update_mediafile_meeting_id.py:44]: This is more or less redundant with the other events (besides the `delete`). What I'm still missing is a list update event (see e.g. https://github.com/OpenSlides/openslides-backend/blob/49a608ff81bc22a956bb8f17d6d157c9fe443f75/migrations/tests/test_0018_remove_template_meeting_committee_relation.py#L37) and a delete fields event (by providing `None` as a value for a field)

File: openslides_backend/action/actions/mediafile/mixins.py [hunk 34]
```diff
@@ -3,30 +3,25 @@
...
-    def check_permissions(self, instance: Dict[str, Any]) -> None:
+    def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
         collection, id_ = self.get_owner_data(instance)
-
         self.check_parent_is_dir_and_owner(
             instance.get("parent_id"), str(instance.get("owner_id"))
         )
         parent_id = instance.get("parent_id")
-        if not parent_id and instance.get("id"):
+        if self.name == "mediafile.update":
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/mixins.py:24]: Mmh, I liked the more generic approach better (`if not parent_id`), so the mixin does not need to know which classes inherit from it.

File: openslides_backend/action/util/assert_belongs_to_meeting.py [hunk 35]
```diff
@@ -26,6 +26,12 @@ def assert_belongs_to_meeting(
             )
             if meeting_id not in instance.get("meeting_ids", []):
                 errors.add(str(fqid))
+        elif fqid.collection.collection == "mediafile":
+            mediafile = datastore.get(fqid, ["owner_id"])
+            collection, id_ = mediafile["owner_id"].split(KEYSEPARATOR)
+            if collection == "meeting":
```
Comment (jsangmeister) [inline openslides_backend/action/util/assert_belongs_to_meeting.py:32]: if the collection is not meeting, an error must also be thrown, since it does not belong to the meeting

File: openslides_backend/action/actions/mediafile/mixins.py [hunk 36]
```diff
@@ -109,8 +109,14 @@ def check_title_parent_unique(
                 filter_ = And(filter_, Not(FilterOperator("id", "=", id_)))
             results = self.datastore.filter(self.model.collection, filter_, ["id"])
             if results:
+                parent_title = "NONE"
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/mixins.py:112]: Confusing for the user. Instead, write something like `File '{title}' already exists in the root folder` in case of no parent

File: openslides_backend/action/actions/mediafile/mixins.py [hunk 37]
```diff
@@ -128,7 +134,10 @@ def check_access_groups_and_owner(
 
     def check_token_unique(self, token: Optional[str], id_: Optional[int]) -> None:
         if token:
-            filter_: Filter = FilterOperator("token", "=", token)
+            filter_: Filter = And(
+                FilterOperator("token", "=", token),
+                FilterOperator("owner_id", "=", "organization/1"),
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/mixins.py:139]: use `KEYSEPARATOR` here

File: openslides_backend/action/actions/mediafile/move.py [hunk 38]
```diff
@@ -78,10 +78,11 @@ def prepare_move_data(
...
         for id_ in ids:
-            if id_ not in db_instances or db_instances[id_].get("owner_id") != owner_id:
-                raise ActionException(f"Id {id_} not in db_instances.")
-
-            instance: Dict[str, Any] = {"id": id_, "parent_id": parent_id}
+            instance: Dict[str, Any] = {
+                "id": id_,
+                "parent_id": parent_id,
+                "owner_id": owner_id,
+            }
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/move.py:85]: Why did you change this? Is it now checked anywhere that no additional instances (which are not in the db) can be given?

File: openslides_backend/action/actions/mediafile/upload.py [hunk 39]
```diff
@@ -55,7 +55,10 @@ def get_updated_instances(self, action_data: ActionData) -> ActionData:
             tokens.append(instance.get("token"))
             results = self.datastore.filter(
                 self.model.collection,
-                FilterOperator("token", "=", instance["token"]),
+                And(
+                    FilterOperator("token", "=", instance["token"]),
+                    FilterOperator("owner_id", "=", "organization/1"),
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/upload.py:60]: use `KEYSEPARATOR`

File: openslides_backend/action/actions/mediafile/upload.py [hunk 40]
```diff
@@ -93,30 +88,24 @@ def update_organization_instance(self, instance: Dict[str, Any]) -> Dict[str, An
...
-        if instance["mimetype"] is None:
-            raise ActionException(f"Cannot guess mimetype for {instance['filename']}.")
-        decoded_file = base64.b64decode(instance["file"])
-        instance["filesize"] = len(decoded_file)
         if instance["mimetype"] == "application/pdf":
             instance["pdf_information"] = self.get_pdf_information(decoded_file)
+        self.media.upload_mediafile(file_, id_, mimetype_)
+        collection, _ = self.get_owner_data(instance)
+        if collection == "meeting":
+            instance["filename"] = filename_
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/upload.py:96]: Orga mediafiles must also have a filename, just remove the `pop` above

File: openslides_backend/action/actions/mediafile/upload.py [hunk 41]
```diff
@@ -93,30 +88,24 @@ def update_organization_instance(self, instance: Dict[str, Any]) -> Dict[str, An
...
+            instance = self.update_access_fields(instance)
```
Comment (jsangmeister) [inline openslides_backend/action/actions/mediafile/upload.py:97]: Do this before uploading the mediafile so you save time and storage in case of an error.

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/mediafile/mixins.py (hunk 34 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/mixins.py:24]: Yes, but the generic version doesn't work, as instance always has an id.

File: openslides_backend/action/actions/mediafile/move.py (hunk 38 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/move.py:85]: The check permission code need the owner id , so it can check the parent/owner_id vs id/owner_id.

File: openslides_backend/action/actions/mediafile/upload.py (hunk 41 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/upload.py:97]: Moved the media service upload at the end of the update_instance method.

File: openslides_backend/action/actions/mediafile/upload.py (hunk 40 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/upload.py:96]: remove the pop of filename and update tests.

File: openslides_backend/action/actions/mediafile/mixins.py (hunk 37 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/mixins.py:139]: Done.

File: openslides_backend/action/actions/mediafile/upload.py (hunk 39 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/upload.py:60]: Done

File: openslides_backend/action/actions/mediafile/mixins.py (hunk 36 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/mixins.py]: Use another error message in parent_id None case. This simplifies the logic.

File: openslides_backend/action/util/assert_belongs_to_meeting.py (hunk 35 above)
Comment (reiterl) [reply openslides_backend/action/util/assert_belongs_to_meeting.py:32]: If the collection is not a meeting collection, throw an error.

File: migrations/tests/test_0020_update_mediafile_meeting_id.py (hunk 33 above)
Comment (reiterl) [reply migrations/tests/test_0020_update_mediafile_meeting_id.py]: Add a list update and two field delete events for the test. I don't think it is important, as the required mediafile/meeting_id is changed into the mediafile/owner_id.

File: openslides_backend/action/actions/mediafile/mixins.py (hunk 34 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/mixins.py]: Use a more generic approach for this issue. Replace the "mediafile.update".

File: openslides_backend/action/actions/mediafile/create_directory.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (jsangmeister) [reply openslides_backend/action/actions/mediafile/create_directory.py:55]: merged now, so you can rebase ind incoorporate this change

File: openslides_backend/action/actions/mediafile/move.py (hunk 38 above)
Comment (jsangmeister) [reply openslides_backend/action/actions/mediafile/move.py:85]: Ok, but you removed the check that not additional ids which are not in the datastore can be given. Is this checked anywhere else? If not, add it back (and also add a test for this, since seemingly it did not trigger an error anywhere).

**Review by jsangmeister — CHANGES_REQUESTED** [review]
Some comments from older reviews still open (look for unresolved conversations). Rest looks good

File: openslides_backend/action/actions/mediafile/create_directory.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/create_directory.py:55]: I have rebased this yesterday.

File: openslides_backend/presenter/check_mediafile_id.py (hunk 22 above)
Comment (reiterl) [reply openslides_backend/presenter/check_mediafile_id.py]: Okay, I have renamed `owner_id_int` into `owner_id`.

File: openslides_backend/action/actions/mediafile/move.py (hunk 38 above)
Comment (reiterl) [reply openslides_backend/action/actions/mediafile/move.py:85]: Ahh, yes, I add the db_instance check again, the owner_id check is done somewhere else.

**Review by jsangmeister — APPROVED** [review]
As discussed: last change will be done in a new issue
